crates/lockfile/src/resolution.rs

```diff
@@ -0,0 +1,250 @@
+            LockfileResolution::Directory(_) | LockfileResolution::Git(_) => None,
+        }
+    }
+}
+
+/// Intermediate helper type for serde.
+#[derive(Deserialize, Serialize, From, TryInto)]
+#[serde(tag = "type", rename_all = "camelCase")]
+enum TaggedResolution {
+    Directory(DirectoryResolution),
+    Git(GitResolution),
+}
+
+/// Intermediate helper type for serde.
+#[derive(Deserialize, Serialize, From, TryInto)]
+#[serde(untagged)]
+enum ResolutionSerde {
+    Tarball(TarballResolution),
+    Registry(RegistryResolution),
+    Tagged(TaggedResolution),
+}
+
+impl From<ResolutionSerde> for LockfileResolution {
+    fn from(value: ResolutionSerde) -> Self {
+        match value {
+            ResolutionSerde::Tarball(resolution) => resolution.into(),
+            ResolutionSerde::Registry(resolution) => resolution.into(),
+            ResolutionSerde::Tagged(TaggedResolution::Directory(resolution)) => resolution.into(),
+            ResolutionSerde::Tagged(TaggedResolution::Git(resolution)) => resolution.into(),
+        }
+    }
+}
+
+impl From<LockfileResolution> for ResolutionSerde {
+    fn from(value: LockfileResolution) -> Self {
+        match value {
+            LockfileResolution::Tarball(resolution) => resolution.into(),
+            LockfileResolution::Registry(resolution) => resolution.into(),
+            LockfileResolution::Directory(resolution) => {
+                resolution.pipe(TaggedResolution::from).into()
+            }
+            LockfileResolution::Git(resolution) => resolution.pipe(TaggedResolution::from).into(),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use pretty_assertions::assert_eq;
+    use text_block_macros::text_block;
+
+    #[test]
+    fn deserialize_tarball_resolution() {
+        eprintln!("CASE: without integrity");
+        let yaml = text_block! {
+            "tarball: file:ts-pipe-compose-0.2.1.tgz"
+        };
+        let received: LockfileResolution = serde_yaml::from_str(yaml).unwrap();
+        dbg!(&received);
+        let expected = LockfileResolution::Tarball(TarballResolution {
+            tarball: "file:ts-pipe-compose-0.2.1.tgz".to_string(),
+            integrity: None,
+        });
+        assert_eq!(received, expected);
+
+        eprintln!("CASE: with integrity");
+        let yaml = text_block! {
+            "tarball: file:ts-pipe-compose-0.2.1.tgz"
+            "integrity: sha512-gf6ZldcfCDyNXPRiW3lQjEP1Z9rrUM/4Cn7BZbv3SdTA82zxWRP8OmLwvGR974uuENhGCFgFdN11z3n1Ofpprg=="
+        };
+        let received: LockfileResolution = serde_yaml::from_str(yaml).unwrap();
+        dbg!(&received);
+        let expected = LockfileResolution::Tarball(TarballResolution {
+            tarball: "file:ts-pipe-compose-0.2.1.tgz".to_string(),
+            integrity: "sha512-gf6ZldcfCDyNXPRiW3lQjEP1Z9rrUM/4Cn7BZbv3SdTA82zxWRP8OmLwvGR974uuENhGCFgFdN11z3n1Ofpprg==".to_string().into()
+        });
+        assert_eq!(received, expected);
```

Comment:
We should use snapshot testing in here. It's extremely unreadable at the moment.

Reply:
Snapshot is good for big objects (such as AST of a programming language for example), but `expected` here is small. Do you think moving the content of the object out of this file a good idea? I personally prefer keeping the small object in the same location as the test code.